barretenberg/cpp/pil/vm2/alu.pil

```diff
@@ -79,84 +124,173 @@ op_id = sel_op_add * constants.AVM_EXEC_OP_ID_ALU_ADD
-
-// Tag errors currently have cases:
-// 1. Input tagged as a field for NOT, DIV, or shift operations or non-field for FDIV operation
-// 2. Mismatched tags for inputs a and b for all opcodes apart from TRUNC
-// 1 is handled by checking FF_TAG_ERR in TAG_ERR_CHECK and 2 is handled in AB_TAGS_CHECK.
-pol FF_TAG_ERR = (sel_op_div + sel_op_not + sel_shift_ops) * sel_is_ff + sel_op_fdiv * IS_NOT_FF;
-pol commit sel_ab_tag_mismatch;
+sel * (TAG_U128_DIFF * (sel_is_u128 * (1 - tag_u128_diff_inv) + tag_u128_diff_inv) + sel_is_u128 - 1) = 0;
+
+/******************************************************************************
+* Error Handling
+******************************************************************************/
+pol commit sel_err; // @boolean Follows from definition (see #[ERR_CHECK]).
+pol commit sel_tag_err; // @boolean Follows from definition (see #[TAG_ERR_CHECK]).
+pol commit sel_ab_tag_mismatch; // @boolean
+pol commit sel_div_0_err; // @boolean
+
+/**
+* sel_err is the main consolidated error selector.
+* sel_err = sel_tag_err || sel_div_0_err
+* sel_tag_err is the consolidated tag error selector.
+* sel_tag_err = sel_ab_tag_mismatch || FF_TAG_ERR
+* Three base error cases:
+* 1) FF_TAG_ERR: Input tagged as a field for NOT, DIV, SHL, SHR or non-field for FDIV operation
+* 2) sel_ab_tag_mismatch: Mismatched tags for inputs a and b for all opcodes apart from TRUNCATE.
+* 3) sel_div_0_err: occurs when DIV or FDIV is performed and b == 0.
+*/
+
+// Boolean conditions
 sel_ab_tag_mismatch * (1 - sel_ab_tag_mismatch) = 0;
+sel_div_0_err * (1 - sel_div_0_err) = 0;
+
+// sel_tag_err and sel_div_0_err can happen simultaneously for DIV and FDIV, so we need to OR them together.
+#[ERR_CHECK]
+sel_err = sel_tag_err + sel_div_0_err - sel_tag_err * sel_div_0_err;
 
-// TODO(MW): It's technically possible to have BOTH cases be true if we perform a DIV or shift with FF ib and integer ia,
-// so for now I take sel_ab_tag_mismatch * FF_TAG_ERR.
+// sel_ab_tag_mismatch and FF_TAG_ERR can happen simultaneously for DIV, SHL, SHR, so we need to OR them together.
 #[TAG_ERR_CHECK]
 sel_tag_err = sel_ab_tag_mismatch + FF_TAG_ERR - sel_ab_tag_mismatch * FF_TAG_ERR;
 
+// ------------------------------------------------------------
+// Error case 1: FF_TAG_ERR
+// ------------------------------------------------------------
+pol FF_TAG_ERR = (sel_op_div + sel_op_not + SHIFT_OPS) * sel_is_ff + sel_op_fdiv * IS_NOT_FF;
+
+// ------------------------------------------------------------
+// Error case 2: sel_ab_tag_mismatch
+// ------------------------------------------------------------
 // For NOT opcode, an error occurs if the tag of a is FF. In this case, tracegen will set
-// b's tag as 0 which, while it would currently pass the checks below, is not a tag inequality we
-// want to throw with sel_ab_tag_mismatch:
-pol CHECK_AB_TAGS = 1 - sel_op_not * sel_is_ff - sel_op_truncate;
+// b's tag as 0 (FF) which, in turn, will cause sel_ab_tag_mismatch to be 0.
```

Comment:
As a separate thing (not introduced by this PR!), I think we should think about or at least point out that a lot of this complication is because we have arbritrarily set the `FF` tag value to be 0 (IIRC). It's not actually because the value is `FF`, just the use of 0 i.e. an empty value.